src/librbd/operation/RenameRequest.cc

```diff
@@ -102,7 +104,13 @@ int RenameRequest<I>::filter_state_return_code(int r) {
   I &image_ctx = this->m_image_ctx;
   CephContext *cct = image_ctx.cct;
 
-  if (m_state == STATE_REMOVE_SOURCE_HEADER && r < 0) {
+  if (m_state == STATE_READ_SOURCE_HEADER && r == -ENOENT) {
+    RWLock::RLocker snap_locker(image_ctx.snap_lock);
+    if (image_ctx.name == m_dest_name) {
+      // signal that replay raced with itself
+      return -EEXIST;
+    }
+  } else if (m_state == STATE_REMOVE_SOURCE_HEADER && r < 0) {
```

Comment:
@dillaman But it looks to me the duplicate `RenameRequest` will still finish with a non-zero error code? Won't it propagate to the user?

Reply:
The ```-EEXIST``` error code is explicitly filtered out for this case [1]. The sporadic failure was occurring when an ```-ENOENT``` was bubbling up if multiple renames were executed concurrently due to a messenger disconnect/reconnect.

[1] https://github.com/ceph/ceph/blob/master/src/librbd/Operations.cc#L537

Reply:
got it. Thanks!